### ISSUE TITLE
Application Error when attempting thebe execution

### ISSUE THREAD

**Issue by ryanlovett** [executablebooks/mystmd#816]
We are not able to execute code-cell blocks with a local jupyter kernel.

## Description

We have a project with the following in `myst.yml`:
```yaml
project:
  thebe:
    server:
      url: http://localhost:8888
      token: test-secret
```

We invoke jupyter lab as
```bash
$ jupyter kernelspec list
Available kernels:
  ir         /path/to/miniforge3/share/jupyter/kernels/ir
  python3    /path/to/miniforge3/share/jupyter/kernels/python3
$ jupyter lab --NotebookApp.token=test-secret --NotebookApp.allow_origin='https://localhost:3000'
```

A file contains:
```yaml
:::{code-cell} python3
import numpy as np
mean = np.mean(np.random.normal(size=100))
print(f"{mean=}")
:::
```

I can run `myst start` and visit the server and load the example file. When I click on the power button:

![Screenshot 2023-12-18 at 5 27 30 PM](https://github.com/executablebooks/mystmd/assets/10248125/2137d3db-74b5-4554-a108-49fdf9548df0)

 to start the kernel, myst directs me to a page with:
```python
TypeError: Cannot read properties of undefined (reading 'reduce')
    at http://localhost:3000/myst_assets_folder/_shared/chunk-CDALZYYP.js:319:27535
    at Array.map (<anonymous>)
    at k5t (http://localhost:3000/myst_assets_folder/_shared/chunk-CDALZYYP.js:319:27053)
    at http://localhost:3000/myst_assets_folder/_shared/chunk-CDALZYYP.js:320:791
    at Ei (http://localhost:3000/myst_assets_folder/_shared/chunk-OXM6VUU4.js:8:24198)
    at yr (http://localhost:3000/myst_assets_folder/_shared/chunk-OXM6VUU4.js:8:42223)
    at http://localhost:3000/myst_assets_folder/_shared/chunk-OXM6VUU4.js:8:40571
    at Ca (http://localhost:3000/myst_assets_folder/_shared/chunk-OXM6VUU4.js:1:62053)
    at MessagePort.Ra (http://localhost:3000/myst_assets_folder/_shared/chunk-OXM6VUU4.js:1:62444)
```

I ran `myst clean --all` and replaced theme.server with theme.lite=true, and then ran`myst init`. I received a similar Application Error.

```bash
$ myst -v
v1.1.36
```

When I pass `-d`, nothing happens after pressing the power button. There is no more additional logging.

If I shutdown my jupyter server, there is no difference in behavior. It doesn't seem like anything is trying to connect to it.

**Comment by stevejpurves** [comment]
@ryanlovett thanks for reporting this! I think this the same as #748 which we should be able to get fixed in the next couple of days

**Comment by stevejpurves** [comment]
@ryanlovett we've made a fix that means text based notebooks are not working, although your error message is different from what I was seeing prior to the fix I'm still expecting that to have addressed this issue.

It would be great if you could install the latest `mystmd` and confirm it's working in your case too.

**Comment by ryanlovett** [comment]
Thanks @stevejpurves ! I upgraded to 1.1.37 via conda-forge, and then created a stripped down example containing:
```markdown
---
title: "Example: Jupyter"
---

This is the mean of some random numbers using numpy.

:::{code-cell} python3
import numpy as np
mean = np.mean(np.random.normal(size=100))
print(f"{mean=}")
:::
```

with the relevant myst.yml settings above. I also ran `myst clean --all` and `myst init`. Unfortunately I'm not seeing the "power" button to turn on the kernel.

Previously the markdown file I was testing had (perhaps erroneously) the following at the top:
```yaml
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
```

That seems to be only relevant for markdown-based notebooks which this file is not. However, when I do re-add that to this markdown file, the power button reappears. Pressing it causes a new error message to appear at the bottom right of the page, containing:

> ⛔️ Error connecting to Local Server ⛔️
> error - Server not reachable (http://localhost:8888/) - TypeError: Failed to fetch

So I cannot reproduce the original exception, but there does not seem to be communication between myst and jupyter. I'm happy to close this issue and open a new one, retitle this one, and/or just continue discussion here.

**Comment by stevejpurves** [comment]
Hi @ryanlovett - ok, so at the moment in order to use the `code-cell` directive and have this behave as a executable cell, the md file does need to be interpreted as a markdown-based notebook (if there's a different use case that we need to be considering for execution in just a myst markdown file without that being a notebook, or having linked content from a notebook then that's worth opening a separate issue on).

So, inserting the jupytext frontmatter, and with the following in your `myst.yml`

```
project:
  thebe:
    server:
      url: http://localhost:8888
      token: test-secret
```

That should be working to debug further, do you have any messages from jupyter i.e. in the terminal where jupyter was started? e.g. a 401 or 405?

Could you also try just temporarily starting jupyter with:

```
jupyter lab --NotebookApp.token=test-secret --NotebookApp.allow_origin='*'
```

In case the issue is with the `allow_origin` header string?

**Comment by ryanlovett** [comment]
Ahhh, I had thought one could integrate executable content into markdown files [similar to quarto](https://quarto.org/docs/computations/execution-options.html). I have read through Integrating [Jupyter based Computation](https://mystmd.org/guide/integrating-jupyter) a few times and yet I still missed the explicit Attention callout (read this, it's important!) which states, "Integrated Jupyter Computation is hot off the press and currently limited to pages that represent complete Jupyter notebooks." I may open a separate ticket for that. While I think it is advantageous to author documents from jupyter, we did get some feedback from one or two people who want to author with simple text editors, and being able to integrate `code-cell` blocks that get evaluated by a kernel would help them.

By "markdown-based notebook" do you mean an .ipynb file with a lot of markdown content? I just read about what jupytext actually is and it appears to be a tool that can parse markdown (or .py) files as if they were real notebooks. This would satisfy those authors who want to use their preferred IDEs, but now I'm more confused about whether "complete Jupyter notebooks" (in the callout) means .ipynb files, *and* .md files that are parseable by jupytext.

The jupytext README says,
> If your notebook is documentation-oriented, a [Markdown-based format](https://github.com/mwouts/jupytext/blob/main/docs/formats-markdown.md) (text notebooks with a .md extension) might be more appropriate. Depending on what you plan to do with your notebook, you might prefer the Myst Markdown format...

So my guess is that code-cells only work in .ipynb files in markdown cells, and they're not supported in .md files, even if they are parseable by jupytext?

I'll follow up on your other questions in a new comment.

**Comment by ryanlovett** [comment]
Re-adding the jupytext section to the frontmatter, and changing `--NotebookApp.allow_origin` from `'https://localhost:3000'` to `*` worked!

I tried restarting jupyter and changed allow_origin back to the localhost URL, restarted myst as `myst start`, force reloaded the localhost:3000 URL in my browser, then tried starting the kernel from the myst UI, and it produced the error in the original report.

That this all did work in a MyST markdown file is great but I'm now even more confused since the doc is not a "complete jupyter notebook".

**Comment by stevejpurves** [comment]
@ryanlovett I just noticed the `https://` in your `allow_origin` string, probably that needs to be `http://`

**Comment by stevejpurves** [comment]
@ryanlovett on the other points you made -- let me try to better explain things in terms of how they are working now. Then we can get to some better documentation (e.g. "complete notebook" is not meaningful).

`mystmd` currently recognises two different files as "notebooks"

1. a markdown file with the `jupytext` or `kernelspec` fields in the frontmatter section (a markdown based notebook)
2. an ipynb file (a json based notebook)

I see your example file with some markdown content and `code-cell` directives as an example of 1. A distinction with `myst` and maybe compared to `quarto` is in order for the markdown file to be executable it currently has to have a `kernelspec` or `jupytext` field in it's frontmatter and the `code-cell` directive doesn't have some of the options over how it is displayed that `quarto` has.

So in cases 1. and 2. the power button and execute controls will appear on the page when jupyter is enabled.

I'm interested to know if that makes sense?

In addition to the above, `myst` has another way that executable content can be included in a markdown file and that is by cross reference - that is explained in https://mystmd.org/guide/reuse-jupyter-outputs where in any markdown file you can include an output (or code, or code+output) from another notebook and that single cell will be available in the markdown file and will be executable. 

When executed, the entire notebook it originates from will be executed up to and including that cell. 

The purpose here is to allow a manuscript to be written that includes tables, figures and generally jupyter outputs that might be created in multiple notebooks. 

This is intended to serve the case where someone has a working set of notebooks like:

* data-cleaning.ipynb
* processing.ipynb
* explore-results.ipynb
* create-figures.ipynb

and then a file `manuscript.md` contains their "article" which includes some cells from those as figures. myst will make those figures (referenced cells) executable in the manuscript, without any code-cell appearing in the `manuscript.md` file.

Again interested to know if that makes sense?

Beyond this, perhaps if `mystmd` was modified to not strictly require `jupytext` or `kernelspec` frontmatter fields, but regard any markdown file with `code-cell` as a notional "notebook" and so by default executable (with a default python kernel) then that would also have removed some of this confusion? if so then we should open a fresh issue for that.

**Comment by stevejpurves** [comment]
additionally
> By "markdown-based notebook" do you mean an .ipynb file with a lot of markdown content? I just read about what jupytext actually is and it appears to be a tool that can parse markdown (or .py) files as if they were real notebooks. This would satisfy those authors who want to use their preferred IDEs, but now I'm more confused about whether "complete Jupyter notebooks" (in the callout) means .ipynb files, and .md files that are parseable by jupytext.

it should mean `.md files that are parseable by jupytext`

**Comment by ryanlovett** [comment]
That's all very helpful @stevejpurves , thanks very much!

Re:
> Beyond this, perhaps if `mystmd` was modified to not strictly require `jupytext` or `kernelspec` frontmatter fields, but regard any markdown file with code-cell as by default executable (so notionally a notebook with a default python kernel) then that would also have removed some of this confusion? if so then we should open a fresh issue for that.

I did expect that all code-cell blocks would be automatically be executed, but I also think that users should be expected to provide information about which kernel they want. If there is to be a default kernel, I think users should be expected to specify what it is, e.g. in the project settings or document frontmatter. In this example file, I specified `:::{code-cell} python3` and it makes logical sense for the kernel selection to be in the `{code-cell}` block and not up top. I hope that myst will be able to run different kernels on a single page. The `kernelspec` field structure, and where it must be specified suggests that only one kernel per page is permitted.

I did not expect that `jupytext` was required for kernel execution, and the values in
```yaml
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
```
don't seem to be related to code execution. (as someone unfamiliar with jupytext)

I do remember being unclear on how the parameter after `{code-cell}` relates to the `kernelspec` field's `.display_name`, `.language`, and `.name` parameters. I specified `:::{code-cell} python3` because `python3` was one of the values produced when I run `jupyter kernelspec list`. This is logical to me, though maybe it is problematic for reproducibility. For example if another person wants to re-execute the same document on their device, they may have named their kernels differently, i.e. perhaps their kernel is named `python` and they might not want to have to rename their kernel or change all of the code-cell parameters. In which case there would have to be a mapping between the kernels that are specified in the markdown files and the those available to jupyter. Is that what `kernelspec` is for? The mapping I imagine could exist in the project settings.

Given something like:
```yaml
project:
  execute:
    code_cells:
     - name: python
       jupyter_kernel: python3
     - name: r
       jupyter_kernel: ir
```
the author could specify `:::{code-cell} r` and it would run their `ir` jupyter kernel. Someone else who wanted to reproduce could then easily override with their own code-cell-to-kernel mapping.

Perhaps what I describe adds too much complexity. Anyways, I realize that my expectations can be different than others', or just not logical or feasible.

To close out this issue, I'll try to make a PR for `https` -> `http` change.
